Student: Duncan Ferguson <br>
Student Id: 871641260 <br>
Class: Comp 4431-1 <br>
Assignment: Assignment 2 <br>
Date: 11/12/2021 <br>
*Project was Completed by Individual, group is just listed below for reference to past code that might
be reused from previous Excercises* <br>
Group: Name: Broken Toe <br>
Group Members: Emma Bright, Mike Santoro <br>

### Assignment Description:
In this assignment you are to use Decision trees and Naive Bayes for classification. In addition, you are to explore
 your models to determine which feature vector attributes you can remove without "significantly hurting"
 classification importance. There is no single "right" answer for this question. Also, "significantly hurting"
 is a judgement call, please explain why you mean by that and provided numeric values. You should start with
 the full dataset and look at the decision tree features_importances_ and naive bayes permutation_importance
 to guid which attributes to remove. You will then want to iteratively remove attributes, possibly one at a time or
 several at a time. Again, there is no single "right way" to do this.

The goal is to predict whether it will rain tomorrow given the days weather statistics. The last column,
"Rain Tomorrow" is the classification value for I have cleaned the data set for you. I removed a few attributes that
were almost all zeros, and replaced missing values with the mean value for that column. Also, strings have been
transformed into numeric values so the data set will work well for both decision trees and naive bayes. <br>

#### What To Turn In: <br>
1). A report that explains which attributes you have eliminated, why those, and what effect this has on the quality of
predictions compared to using the full dataset. Feel free to include a table of feature_importances and
 permutation_importances that show how any why you reduced the set of attributes. <br>

2). Your code for creating the decision tree and naive bayes models and getting the feature/permuttion importances.
Your code does not need to cover your whole experimental space, but should provide enough information to show how you
made at least one decision about which attribute (column) to eliminate


In [648]:
# Importing main python packages for assignment
import numpy as np
import pandas as pd
from sklearn import tree
from sklearn.naive_bayes import *
from sklearn.model_selection import train_test_split
from sklearn.metrics import *
from sklearn.inspection import permutation_importance

In [649]:
def train_data(df, drops=None):
    """This function trains the data and returns the desired drops"""
    # Y is the classification
    Y = df['RainTomorrow'].tolist()
    X = df.copy()
    X = X.drop(columns=["RainTomorrow"])
    if drops != None:
        X = X.drop(columns=drops)
    x_train, x_test, y_train, y_test = train_test_split(X, Y, test_size=0.25, random_state=2)
    return x_train, x_test, y_train, y_test, X

In [650]:
def decisionTree_Model(x_train, x_test, y_train, y_test, X):
    """This Function Goes through and prints out the Decision tree model"""
    # Building out a decision tree
    dtree = tree.DecisionTreeClassifier(criterion="gini")
    dtree = dtree.fit(x_train, y_train)
    y_predicted = dtree.predict(x_test)
    accuracy = accuracy_score(y_test, y_predicted)
    print("DecisionTree Accuracy: ", accuracy)
    print("DecisionTree Confusion Matrix:")
    print(confusion_matrix(y_test, y_predicted))
    # return importance_df(dtree.feature_importances_, X)
    important = dtree.feature_importances_
    df_importance_list = []
    for i, v in enumerate(important):
        df_importance_list.append([X.columns[i], v])
    df_importance = pd.DataFrame(df_importance_list, columns=["FName", "Score"])
    df_importance.sort_values(by=['Score'], ascending=False, inplace=True)
    print("decision tree dtree feature importance:")
    print(df_importance)
    decisionTree_Model_Features = df_importance["FName"].tolist()
    return accuracy, decisionTree_Model_Features

def decisionTree_Model_NP(x_train, x_test, y_train, y_test, X):
    """This Function Goes through and prints out the Decision tree model without the print functions"""
    # Building out a decision tree
    dtree = tree.DecisionTreeClassifier(criterion="gini")
    dtree = dtree.fit(x_train, y_train)
    y_predicted = dtree.predict(x_test)
    accuracy = accuracy_score(y_test, y_predicted)
    important = dtree.feature_importances_
    df_importance_list = []
    for i, v in enumerate(important):
        df_importance_list.append([X.columns[i], v])
    df_importance = pd.DataFrame(df_importance_list, columns=["FName", "Score"])
    df_importance.sort_values(by=['Score'], ascending=False, inplace=True)
    decisionTree_Model_Features = df_importance["FName"].tolist()
    return accuracy, decisionTree_Model_Features


In [651]:
def GausianNB_Model(x_train, x_test, y_train, y_test, X):
    """This Function takes in the training and the testing models to help create the GausianNB Model"""
    model2 = GaussianNB()
    model2.fit(x_train, y_train)
    gausianNB_predicted = model2.predict(x_test)
    print('\nconfusion_matrix from Gaussian naive bayes:')
    print(confusion_matrix( y_test, gausianNB_predicted ) )
    accuracy = accuracy_score(y_test, gausianNB_predicted)
    print('accuracy = ' + str(accuracy))
    imps = permutation_importance(model2, x_test, y_test)
    print("gausianNB feature importance:")
    df_Gausian_Feature_Importance_List = []
    Gausian_Feature_Importance_List =  imps.importances_mean.tolist()
    for row in enumerate(Gausian_Feature_Importance_List):
        df_Gausian_Feature_Importance_List.append([X.columns[row[0]], row[1]])
    df_Gausian_Feature_Importance = pd.DataFrame(df_Gausian_Feature_Importance_List, columns=["Feature", "Significance"])
    df_Gausian_Feature_Importance.sort_values(by=["Significance"], ascending=False, inplace=True)
    GausianND_Features = df_Gausian_Feature_Importance["Feature"].tolist()
    print(GausianND_Features)
    print(df_Gausian_Feature_Importance)
    return accuracy, GausianND_Features
    # return df_Gausian_Feature_Importance

In [652]:
np.random.seed(0)
# Importing data
df = pd.read_csv("assignment2_cleanInfile.csv")
x_train, x_test, y_train, y_test, X = train_data(df)

In [653]:
d1 = decisionTree_Model(x_train, x_test, y_train, y_test, X)

DecisionTree Accuracy:  0.7823567470252327
DecisionTree Confusion Matrix:
[[23529  3983]
 [ 3754  4283]]
decision tree dtree feature importance:
            FName     Score
11    Humidity3pm  0.266493
13    Pressure3pm  0.070134
5   WindGustSpeed  0.065499
3        Rainfall  0.061755
1         MinTemp  0.052217
10    Humidity9am  0.052194
12    Pressure9am  0.048039
16        Temp9am  0.042975
2         MaxTemp  0.042970
17        Temp3pm  0.041659
0        Location  0.040381
8    WindSpeed9am  0.036237
9    WindSpeed3pm  0.034998
15       Cloud3pm  0.031621
4     WindGustDir  0.030955
6      WindDir9am  0.030737
7      WindDir3pm  0.030037
14       Cloud9am  0.019316
18      RainToday  0.001781


In [654]:
g1 = GausianNB_Model(x_train, x_test, y_train, y_test, X)


confusion_matrix from Gaussian naive bayes:
[[24207  3305]
 [ 3552  4485]]
accuracy = 0.807111311147993
gausianNB feature importance:
['Humidity3pm', 'Rainfall', 'Pressure3pm', 'WindGustSpeed', 'Pressure9am', 'Cloud3pm', 'MinTemp', 'Cloud9am', 'Humidity9am', 'Location', 'WindSpeed9am', 'WindSpeed3pm', 'Temp9am', 'WindDir3pm', 'WindGustDir', 'RainToday', 'WindDir9am', 'Temp3pm', 'MaxTemp']
          Feature  Significance
11    Humidity3pm      0.016034
3        Rainfall      0.012445
13    Pressure3pm      0.004737
5   WindGustSpeed      0.004670
12    Pressure9am      0.003061
15       Cloud3pm      0.002712
1         MinTemp      0.000805
14       Cloud9am      0.000428
10    Humidity9am      0.000084
0        Location     -0.000056
8    WindSpeed9am     -0.000264
9    WindSpeed3pm     -0.000428
16        Temp9am     -0.000467
7      WindDir3pm     -0.000473
4     WindGustDir     -0.000776
18      RainToday     -0.001046
6      WindDir9am     -0.001204
17        Temp3pm     -0.001328

Looking at the first model assumptions we can definitely conclude that the Humidity3pm column is the most important
for determining a significance. But there are a few more ways to go through this. The example below will show the
new models when dropping "RainToday" which has the least significance on the decisionTree Model

In [655]:
x_train, x_test, y_train, y_test, X = train_data(df, "RainToday")
d2 = decisionTree_Model(x_train, x_test, y_train, y_test, X)

DecisionTree Accuracy:  0.7802751132240007
DecisionTree Confusion Matrix:
[[23487  4025]
 [ 3786  4251]]
decision tree dtree feature importance:
            FName     Score
11    Humidity3pm  0.266565
13    Pressure3pm  0.070425
5   WindGustSpeed  0.065828
3        Rainfall  0.063135
10    Humidity9am  0.052556
1         MinTemp  0.051678
12    Pressure9am  0.046830
2         MaxTemp  0.043494
17        Temp3pm  0.042539
16        Temp9am  0.041994
0        Location  0.041105
9    WindSpeed3pm  0.036300
8    WindSpeed9am  0.035512
6      WindDir9am  0.031471
4     WindGustDir  0.031423
15       Cloud3pm  0.030239
7      WindDir3pm  0.029796
14       Cloud9am  0.019108


In [656]:
g2 = GausianNB_Model(x_train, x_test, y_train, y_test, X)


confusion_matrix from Gaussian naive bayes:
[[24769  2743]
 [ 3747  4290]]
accuracy = 0.8174350895946441
gausianNB feature importance:
['Humidity3pm', 'Rainfall', 'WindGustSpeed', 'Cloud3pm', 'Pressure3pm', 'Humidity9am', 'Pressure9am', 'MinTemp', 'Cloud9am', 'Location', 'Temp9am', 'WindGustDir', 'WindDir3pm', 'WindDir9am', 'Temp3pm', 'WindSpeed9am', 'WindSpeed3pm', 'MaxTemp']
          Feature  Significance
11    Humidity3pm      0.023455
3        Rainfall      0.016225
5   WindGustSpeed      0.005018
15       Cloud3pm      0.004506
13    Pressure3pm      0.004388
10    Humidity9am      0.003539
12    Pressure9am      0.003151
1         MinTemp      0.001980
14       Cloud9am      0.000422
0        Location      0.000135
16        Temp9am      0.000045
4     WindGustDir     -0.000287
7      WindDir3pm     -0.000366
6      WindDir9am     -0.000405
17        Temp3pm     -0.000534
8    WindSpeed9am     -0.000585
9    WindSpeed3pm     -0.000872
2         MaxTemp     -0.000928


Interestingly enough. When we drop the Rain Today the decision tree model accuracy model goes down.
Where as Gausiann model accuracy goes up.
Decision Tree Accuracy 0.7823567470252327 -> 0.7802751132240007
Gaussian Accuracy 0.807111311147993 -> 0.8174350895946441

Let's first look at the decision tree and go with a forward select. From our first two run's obviously Humidity3pm
has the most significance. So We'll keep that then add the next category that bumps up the value

In [657]:
def forward_selection_DT(accuracy, features, selection, df):
    """Forward Selection Loop"""
    new_accuracy = accuracy
    features_loop = [word for word in features if word not in selection]
    for feature in features_loop:
        droplist = [word for word in features if word != feature]
        x_train, x_test, y_train, y_test, X = train_data(df, droplist)
        d_loop = decisionTree_Model_NP(x_train, x_test, y_train, y_test, X)
        if d_loop[0] >= new_accuracy:
            new_accuracy = d_loop[0]
            selection = selection + d_loop[1]
    return new_accuracy, selection


features = df.columns.tolist()
features.pop(-1)
accuracy = 0
selection = []
forward_selection_DT(accuracy, features, selection, df)

(0.8204168893639765, ['Location', 'MinTemp', 'Rainfall', 'Humidity3pm'])

This Forward Selection model on the Decision tree leaves us with Location, minTemp, rainfall, and huimidity as the
best predictors.

In [658]:
selection = ['Location', 'MinTemp', 'Rainfall', 'Humidity3pm']
new_drop = [word for word in features if word not in selection]
x_train, x_test, y_train, y_test, X = train_data(df, new_drop)
d4 = decisionTree_Model(x_train, x_test, y_train, y_test, X)

DecisionTree Accuracy:  0.7549016849981716
DecisionTree Confusion Matrix:
[[23149  4363]
 [ 4350  3687]]
decision tree dtree feature importance:
         FName     Score
3  Humidity3pm  0.372432
1      MinTemp  0.328253
0     Location  0.163198
2     Rainfall  0.136116
